lib/client/ca_export.go

```diff
@@ -0,0 +1,289 @@
+// that can be placed in ~/.ssh/authorized_hosts. The format adheres to the man sshd (8)
+// authorized_hosts format, a space-separated list of: marker, hosts, key, and comment.
+// For example:
+// > @cert-authority *.cluster-a ssh-rsa AAA... type=host
+// URL encoding is used to pass the CA type and allowed logins into the comment field.
+func ExportAuthorities(ctx context.Context, client auth.ClientI, req ExportAuthoritiesRequest) (string, error) {
+	var typesToExport []types.CertAuthType
+
+	// this means to export TLS authority
+	switch req.AuthType {
+	// "tls" is supported for backwards compatibility.
+	// "tls-host" and "tls-user" were added later to allow export of the user
+	// TLS CA.
+	case "tls", "tls-host":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.HostCA,
+			UnpackPEM:         false,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	case "tls-user":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.UserCA,
+			UnpackPEM:         false,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	case "db":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.DatabaseCA,
+			UnpackPEM:         false,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	case "tls-user-der", "windows":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.UserCA,
+			UnpackPEM:         true,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	}
+
+	// If no AuthType is given, export both HostCA and UserCA.
+	if req.AuthType == "" {
+		typesToExport = []types.CertAuthType{types.HostCA, types.UserCA}
+	} else {
+		authType := types.CertAuthType(req.AuthType)
+		if err := authType.Check(); err != nil {
+			return "", trace.Wrap(err)
+		}
+		typesToExport = []types.CertAuthType{authType}
+	}
+	localAuthName, err := client.GetDomainName(ctx)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// fetch authorities via auth API (and only take local CAs, ignoring
+	// trusted ones)
+	var authorities []types.CertAuthority
+	for _, at := range typesToExport {
+		cas, err := client.GetCertAuthorities(ctx, at, req.ExportPrivateKeys)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		for _, ca := range cas {
+			if ca.GetClusterName() == localAuthName {
+				authorities = append(authorities, ca)
+			}
+		}
+	}
+
+	ret := strings.Builder{}
+	for _, ca := range authorities {
+		if req.ExportPrivateKeys {
+			for _, key := range ca.GetActiveKeys().SSH {
+				fingerprint, err := sshutils.PrivateKeyFingerprint(key.PrivateKey)
+				if err != nil {
+					return "", trace.Wrap(err)
+				}
+				if req.ExportAuthorityFingerprint != "" && fingerprint != req.ExportAuthorityFingerprint {
+					continue
+				}
+				ret.WriteString(string(key.PrivateKey))
```

Comment:
```suggestion
				ret.Write(key.PrivateKey)
```

(Could lead to less copying.)